modules: lvgl: add automatic lv_timer_handler call

uLipe opened the pull request:
inside of its own thread when user enable it on
the LVGL Zephyr options menu, using this option makes the calling of the lv_timer_handler driven by an internal timer and thread. Options like priority and stack size for this thread are exposed making it configurable by the application.

Worth to mention, this option is disbaled by default due the nature of the current samples of the LVGL that invokes lv_timer_handler from the application

uLipe committed:
modules: lvgl: add automatic lv_timer_handler call

inside of its own thread when user enable it on
the LVGL Zephyr options menu, using this option makes
the calling of the lv_timer_handler driven by an internal
timer and thread. Options like priority and stack size
for this thread are exposed making it configurable by the
application.

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>

modules: lvgl: put LVGL core into a dedicated workqueue

Replacing the initial approach based on timers and semaphore

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>

faxe1008 commented on modules/lvgl/Kconfig:
```diff
@@ -153,6 +153,31 @@ config LV_Z_INIT_PRIORITY
 	  Priority used for the automatic initialization of LVGL.
 
+config LV_Z_HANDLER_THREAD
+	bool "Drive lv_timer_handler in a separate thread"
```
Nit: something like "Run LVGL event loop in a seperate thread" more resilient against API renaming.

faxe1008 commented on modules/lvgl/lvgl.c:
```diff
@@ -219,6 +219,34 @@ static int lvgl_allocate_rendering_buffers(lv_display_t *display)
+		k_sem_take(&lv_timer_sem, K_FOREVER);
+		uint32_t reload_value = lv_timer_handler();
+		/*schedule next timer verification*/
+		k_timer_start(&lv_handler_timer, K_MSEC(MIN(reload_value, INT32_MAX)), K_MSEC(0));
```
I suppose the semaphore we still need to ensure that `lvgl_init` is called beforehand, but why would we need the timer? I think having a simple sleep here is essentially the same - allowing another thread to run, also I think it does not give us anything precision wise as well.

uLipe replied:
I was asking the same thing until I compared the sleep loop with timer based approach, the main advantage here is the timing slippage is better compensated on virtual timers, this reflects for example on having rock solid 30ms without perceptible
jitter when running the demo benchmark.

The soft timer usage approach will show up more interesting when running at 60Hz of the refresh rate.

So here we get an extra ROM+RAM overhead in exchange of having a better frame counter :)

uLipe commented on modules/lvgl/Kconfig:
```diff
@@ -153,6 +153,31 @@ config LV_Z_INIT_PRIORITY
 
+config LV_Z_HANDLER_THREAD
+	bool "Drive lv_timer_handler in a separate thread"
+	depends on LV_Z_AUTO_INIT
```
Yes this is more a software pattern than technical challenge that appeared, also by separating the lvgl core on its own thread we ensure there will be nothing else running that may affect the timer_handler.

The danger of having the lv_timer_handler being called from the application is to defer the responsibillity to the user of calling the lvgl core properly while LVGL is a subsystem from Zephyr so driving it should
not ( in some cases) responsibility of the application.

BTW NuttX also offers something similar but it relies on its libuv port.

JarmouniA reviewed (commented):
We will need a testcase somewhere in tests/samples with this option enabled

AndreCostaaa commented on modules/lvgl/lvgl.c:
```diff
@@ -219,6 +219,34 @@ static int lvgl_allocate_rendering_buffers(lv_display_t *display)
+{
+	while (1) {
+		k_sem_take(&lv_timer_sem, K_FOREVER);
+		uint32_t reload_value = lv_timer_handler();
```
Would it make sense to use `CONFIG_LV_DEF_REFR_PERIOD` in case this returns `LV_NO_TIMER_READY` instead of `INT32_MAX` ?

AndreCostaaa reviewed (commented):
Pretty cool feature 🚀

pdgendt reviewed (commented):
A cleaner solution IMO is:

- Have a `k_work_q` as the "thread"
- Use a `k_work_delayable` to be (re)scheduled
- Call `lv_timer_hander` in the delayed work handler and re-schedule based on the result

EDIT: Added benefit: the work queue could be user exposed to allow submitting/scheduling more work onto the LVGL thread.

pdgendt commented:
> @pdgendt I implemented using the workqueus, I kepi the previous one on the commit history, for easy comparisation

Didn't notice this, please squash in that case.

uLipe commented:
@JarmouniA what would be a good way for testing this option? Just duplicate one of the samples but using this option and removing the sleep loop? BTW this can be part of another PR?

uLipe committed:
modules: lvgl: expose the lvgl workqueue to the user application

allowing user to submit their own work intems to it.

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>

pdgendt commented on modules/lvgl/include/lvgl_zephyr.h:
```diff
@@ -32,6 +32,20 @@ extern "C" {
+ *
+ * @return 0 on success, negative errno code on failure
+ */
+int lvgl_schedule_job(int32_t ms);
```
This is not what I meant, but something like this:

```c
struct k_work_q *lvgl_workqueue(void);
```

This would allow users to have their own `k_work` items submitted/scheduled on the LVGL thread, making it thread-safe to call `lv_` functions there.

pdgendt reviewed (dismissed):
Awesome stuff, do you see the same gains you had like using the timer? It probably should as these use the same low level primitives.

uLipe commented:
@pdgendt yes they offer the same performance and better timer slippage correction than using the  sleep loop approach, but the workqueue is much cleaner

uLipe committed:
samples: modules: lvgl: demos: added new test cases

that uses the LVGL internal workqueue option to
drive the LVGL core automatically

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>

uLipe commented:
Added the tests case on the demo samples as pointed by @JarmouniA

faxe1008 reviewed (dismissed):
Agreeing with @JarmouniA, one case for building the sample with the new option should be good enough, otherwise LGTM :^)

uLipe commented:
@JarmouniA @faxe1008 , got it, I left only the case of the demo benchmark then please PTAL again :)

pdgendt commented on samples/modules/lvgl/demos/src/main.c:
If `CONFIG_LV_Z_RUN_LVGL_ON_WORKQUEUE` is enabled the `main` and `lvgl` threads. I'll post a fix PR.
